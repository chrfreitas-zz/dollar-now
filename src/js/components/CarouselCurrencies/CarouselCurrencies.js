import React, { Component } from 'react';
import styled from 'styled-components';

import Controls from './Controls/Controls';
import Items from './Items/Items';

const Container = styled.div`
    position: relative;
    width: 330px;
    height: 75px;
`;

class CarouselCurrencies extends Component {
    state = {
        currencies: [
            {
                code: 'EUR',
                value: '1,18',
                active: false
            },
            {
                code: 'BRL',
                value: '3,25',
                active: true
            },
            {
                code: 'CAD',
                value: '0,95',
                active: false
            }
        ],
        activePosition: -75
    }

    handleChangeCurrency = (moveIndex) => {
        this.setState(prevState => ({
            currencies: this.changeActiveItem(moveIndex, prevState.currencies),
            activePosition: this.changePosition(moveIndex, prevState.activePosition)
        }));
    }

    changeActiveItem(moveIndex, currencies){
        const currentIndexActive = currencies.findIndex(item => item.active);

        return currencies.map((item, index) => {
            const nextIndexActive = currentIndexActive + moveIndex;

            item.active = false;

            if(nextIndexActive === index){
                item.active = true
            }

            return item;
        })
    }

    changePosition(moveIndex, currentPosition) {
        const toDown = Math.sign(moveIndex) === -1;

        if (toDown) {
            return currentPosition + 55;
        }

        return currentPosition - 55
    }

    render() {
        return (
            <Container>
                <Items currencies={this.state.currencies} activePosition={this.state.activePosition}/>
                <Controls handleChangeCurrency={this.handleChangeCurrency} />
            </Container>
        )
    }
}

export default CarouselCurrencies;