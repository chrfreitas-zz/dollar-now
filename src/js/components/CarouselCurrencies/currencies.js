import React, { Component } from 'react';
import styled from 'styled-components';

import Controls from './Controls/Controls';

const Container = styled.div`
    position: relative;
    width: 330px;
    height: 75px;
`;

const CurrenciesContainer = styled.div`
    position: absolute;
    top: 0;
    transition: top .4s;
    transform: translateX(-50%);
    left: 50%;
`;

const Item = styled.div`
    font-size: 30px;
    margin: 20px;
    opacity: .2;
    display: flex;
    flex-direction: row;
    justify-content: center;

    &.active {
        font-size: 60px
        opacity: 1;
        transition: opacity .3s, font-size .3s;
    }
`;

const CurrencyCode = styled.div`
    margin: 0 30px 0 0;
`;



class Currencies extends Component {
    constructor(props) {
        super(props);

        this.state = {
            currencies: [
                {
                    code: 'EUR',
                    value: '1,18',
                    active: false
                },
                {
                    code: 'BRL',
                    value: '3,25',
                    active: true
                },
                {
                    code: 'CAD',
                    value: '0,95',
                    active: false
                }
            ],
            activePosition: -75
        }
    }

    renderList() {
        return this.state.currencies.map((item, index) => (
            <Item className={ item.active ? 'active': '' } key={index}>
                <CurrencyCode>{ item.code }</CurrencyCode>
                <div>{ item.value }</div>
            </Item>
        ));
    }

    handleChangeCurrency(to) {
        const indexActive = this.state.currencies.findIndex(item => item.active);

        this.setState(prevState => ({
            currencies: prevState.currencies.map((item, index) => {
                item.active = false;

                if((indexActive + to)  === index){
                    item.active = true
                }

                return item;
            }),
            activePosition: (Math.sign(to) === -1) ? prevState.activePosition + 55 : prevState.activePosition - 55
        }));
    }

    render() {
        return (
            <Container>
                <CurrenciesContainer style={ {'top': this.state.activePosition} }>
                    { this.renderList() }
                </CurrenciesContainer>
                <Constrols handleChangeCurrency={this.handleChangeCurrency} />
            </Container>
        )
    }
}

export default Currencies;