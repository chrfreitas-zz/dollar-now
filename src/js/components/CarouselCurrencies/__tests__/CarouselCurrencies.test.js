import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import CarouselCurrencies from './../CarouselCurrencies';

beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
});

describe('CarouselCurrencies', () => {
    let wrapper;
    beforeAll(() => {
        wrapper = shallow(<CarouselCurrencies />);
    });

    it('should match with snapshot', () => {
        const snapshot = renderer.create(<CarouselCurrencies />);
        expect(snapshot).toMatchSnapshot();
    });

    it('should return less than 20 when movement is positive', () => {
        expect(wrapper.instance().changePosition(+1, 20)).toBeLessThan(20);
    });

    it('should return greater than 20 when movement is negative', () => {
        expect(wrapper.instance().changePosition(-1, 20)).toBeGreaterThan(20);
    });

    it('should return greater than 20 when movement is negative', () => {
        const currencies = [
            {
                code: 'EUR',
                value: '1,18',
                active: false
            },
            {
                code: 'BRL',
                value: '3,25',
                active: true
            },
            {
                code: 'CAD',
                value: '0,95',
                active: false
            }
        ];

        const expected = [
            {
                code: 'EUR',
                value: '1,18',
                active: false
            },
            {
                code: 'BRL',
                value: '3,25',
                active: false
            },
            {
                code: 'CAD',
                value: '0,95',
                active: true
            }
        ];

        expect(wrapper.instance().changeActiveItem(1, currencies)).toEqual(expected);
    });

    it('shoudl return true to disable button up', () => {
        const currencies = [
            {
                code: 'EUR',
                value: '1,18',
                active: true
            },
            {
                code: 'BRL',
                value: '3,25',
                active: false
            },
            {
                code: 'CAD',
                value: '0,95',
                active: false
            }
        ];

        expect(wrapper.instance().isButtonUpDisabled(-1, currencies)).toBeTruthy();
    });

    it('shoudl return true to disable button down', () => {
        const currencies = [
            {
                code: 'EUR',
                value: '1,18',
                active: false
            },
            {
                code: 'BRL',
                value: '3,25',
                active: false
            },
            {
                code: 'CAD',
                value: '0,95',
                active: true
            }
        ];

        expect(wrapper.instance().isButtonDownDisabled(1, currencies)).toBeTruthy();
    })

});


