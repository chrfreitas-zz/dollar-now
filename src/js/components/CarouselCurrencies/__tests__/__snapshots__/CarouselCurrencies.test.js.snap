// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselCurrencies should match with snapshot 1`] = `
<div
  className="sc-bZQynM jicYgv"
>
  <div
    className="sc-bxivhb enRaSl"
    style={
      Object {
        "top": -75,
      }
    }
  >
    <div
      className="sc-ifAKCX hepNmx"
    >
      <div
        className="sc-EHOje clTNRt"
      >
        EUR
      </div>
      <div>
        1,18
      </div>
    </div>
    <div
      className="active sc-ifAKCX hepNmx"
    >
      <div
        className="sc-EHOje clTNRt"
      >
        BRL
      </div>
      <div>
        3,25
      </div>
    </div>
    <div
      className="sc-ifAKCX hepNmx"
    >
      <div
        className="sc-EHOje clTNRt"
      >
        CAD
      </div>
      <div>
        0,95
      </div>
    </div>
  </div>
  <div
    className="sc-bwzfXH eRJyyo"
  >
    <button
      className="sc-htpNat ZEJQa"
      disabled={false}
      onClick={[Function]}
    >
      <svg
        height="20"
        viewBox="0 0 64 64"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.352 48.296l28.56-28.328 28.58 28.347c.397.394.917.59 1.436.59.52 0 1.04-.196 1.436-.59a2.003 2.003 0 0 0 0-2.849l-29.98-29.735c-.2-.2-.494-.375-.757-.475a2.052 2.052 0 0 0-2.166.456L.479 45.447a2.003 2.003 0 0 0 0 2.849 2.045 2.045 0 0 0 2.873 0z"
          fill="#FFF"
        />
      </svg>
    </button>
    <button
      className="sc-htpNat ZEJQa"
      disabled={false}
      onClick={[Function]}
    >
      <svg
        height="20"
        viewBox="0 0 64 64"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M60.533 15.733l-28.56 28.328L3.394 15.713c-.397-.394-.917-.59-1.437-.59s-1.039.196-1.436.59a2.003 2.003 0 0 0 0 2.849l29.98 29.735c.2.2.494.375.757.476.75.282 1.597.107 2.166-.456l29.981-29.735a2.003 2.003 0 0 0 0-2.849 2.044 2.044 0 0 0-2.872 0z"
          fill="#FFF"
        />
      </svg>
    </button>
  </div>
</div>
`;
